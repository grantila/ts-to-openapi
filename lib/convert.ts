import { createGenerator } from "ts-json-schema-generator"

import { Config } from "./types"
import { getConfig } from "./config"
import { overwriteObject } from "./utils"
import { toYAML } from "./yaml"
import { convertNullToNullable } from "./json-schema"


type SchemaType = ReturnType<
	ReturnType<typeof createGenerator>[ "createSchema" ]
>;

type Definitions = SchemaType[ "definitions" ];

function isExpandable( key: string, expandableTypes: ReadonlyArray< string > )
{
	return expandableTypes.some( type => key.startsWith( `${type}<` ) );
}

function modifySchema( schema: SchemaType, config: Config )
{
	const { expandTypes, removeProps, detectGraphQL, graphQLExpandedTypes } =
		config;

	if ( !schema.definitions )
		return;

	const { definitions } = schema;

	const expandableTypes = Object.keys( definitions )
		.filter( name => isExpandable( name, expandTypes ) );

	const expandName = ( name: string ) =>
		`#/definitions/${encodeURIComponent( name )}`;

	const renamedRef = ( ref: string ) =>
		ref.replace( "#/definitions/", "#/components/schemas/" );

	const expandableTypeMap: { [ type: string ]: any } = expandableTypes
		.reduce(
			( prev, cur ) =>
				Object.assign(
					prev,
					{ [ expandName( cur ) ]: definitions[ cur ] }
				)
			,
			{ }
		);

	// Fix 'null' values for GraphQL Maybe<> types
	if ( detectGraphQL )
	{
		Object.keys( definitions )
			.filter( key => isExpandable( key, graphQLExpandedTypes ) )
			.forEach( key =>
			{
				const type = definitions[ key ] as Record< string, any >;
				convertNullToNullable( type );
			} );
	}

	// Expand types recursively
	const recurse = ( obj: any ) =>
	{
		if ( typeof obj !== "object" )
			return;

		const toRemove: Array< string > = [ ];

		for ( const [ key, value ] of Object.entries( obj ) )
		{
			if ( value && typeof value === "object" )
			{
				if ( Array.isArray( value ) )
					value.forEach( ( value: any ) =>
					{
						recurse( value );
					} );
				else
					recurse( value );
			}

			if ( key === "$ref" && typeof value === "string" )
			{
				if ( expandableTypeMap[ value ] )
				{
					// Replace the object inline with the expanded type
					overwriteObject( obj, expandableTypeMap[ value ] );
					return;
				}
				else
					obj[ key ] = renamedRef( value );
			}

			if ( removeProps.includes( key ) )
				toRemove.push( key );
		}

		toRemove.forEach( key =>
		{
			delete obj[ key ];
		} );
	};
	recurse( definitions );

	expandableTypes.forEach( type =>
	{
		delete definitions[ type ];
	} );
}

export function convert( opts: Partial< Config > )
{
	const { config, schemaGeneratorConfig } = getConfig( opts );

	const generator = createGenerator( schemaGeneratorConfig );

	const definitions: Array<Definitions> = [ ];

	config.types.forEach( type =>
	{
		const schema = generator.createSchema( type );
		modifySchema( schema, config );
		definitions.push( schema.definitions );
	});

	const full = {
		components: {
			schemas: Object.assign( { }, ...definitions )
		}
	};

	return toYAML( full, config );
}
