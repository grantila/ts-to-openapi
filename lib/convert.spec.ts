import { write } from "tempy"

import { convert } from "./convert"

async function convertBlob( tsInterface: string, tsName: string )
{
	const path = await write( tsInterface, { extension: 'ts' } );
	return convert( { path, asComment: false, types: [ tsName ] } );
}

describe( "convert", ( ) =>
{
	it( "simple string type", async ( ) =>
	{
		const yaml = await convertBlob(
			"export type Foo = string;",
			"Foo"
		);
		expect( yaml ).toMatchSnapshot( );
	} );

	// Fixes #2
	describe( "split multiple types into anyOf", ( ) =>
	{
		it( "single top-level", async ( ) =>
		{
			const yaml = await convertBlob(
				"export type Foo = string | number;",
				"Foo"
			);
			expect( yaml ).toMatchSnapshot( );
		} );
		it( "recursively", async ( ) =>
		{
			const yaml = await convertBlob(`
				export type Foo = number | {
					foo: string | boolean;
					bar: string[] | { baz?: number } | { baz: boolean };
				}
				`,
				"Foo"
			);
			expect( yaml ).toMatchSnapshot( );
		} );
	} );
} );
