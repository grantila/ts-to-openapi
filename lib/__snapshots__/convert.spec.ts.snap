// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convert simple string type (JSON) 1`] = `
"{
  \\"openapi\\": \\"3.0.0\\",
  \\"info\\": {
    \\"version\\": \\"1.0.0\\",
    \\"title\\": \\"Foo\\"
  },
  \\"paths\\": {},
  \\"components\\": {
    \\"schemas\\": {
      \\"Foo\\": {
        \\"type\\": \\"string\\"
      }
    }
  }
}"
`;

exports[`convert simple string type 1`] = `
"components:
  schemas:
    Foo:
      type: string
"
`;

exports[`convert split multiple types into anyOf recursively 1`] = `
"components:
  schemas:
    Foo:
      anyOf:
        - type: number
        - additionalProperties: false
          properties:
            bar:
              anyOf:
                - items:
                    type: string
                  type: array
                - additionalProperties: false
                  properties:
                    baz:
                      type: number
                  type: object
                - additionalProperties: false
                  properties:
                    baz:
                      type: boolean
                  required:
                    - baz
                  type: object
            foo:
              anyOf:
                - type: string
                - type: boolean
          required:
            - foo
            - bar
          type: object
"
`;

exports[`convert split multiple types into anyOf single top-level 1`] = `
"components:
  schemas:
    Foo:
      anyOf:
        - type: string
        - type: number
"
`;
